logs_processor.go

```diff
@@ -0,0 +1,239 @@
+	UrlParserType           = "url-parser"
+	UserAgentParserType     = "user-agent-parser"
+)
+
+// LogsProcessor struct represents the processor object from Config API.
+type LogsProcessor struct {
+	Name       *string     `json:"name"`
+	IsEnabled  *bool       `json:"is_enabled"`
+	Type       *string     `json:"type"`
+	Definition interface{} `json:"definition"`
+}
+
+// ArithmeticProcessor struct represents unique part of arithmetic processor
+// object from config API.
+type ArithmeticProcessor struct {
+	Expression       *string `json:"expression"`
+	Target           *string `json:"target"`
+	IsReplaceMissing *bool   `json:"is_replace_missing"`
+}
+
+// AttributeRemapper struct represents unique part of attribute remapper object
+// from config API.
+type AttributeRemapper struct {
+	Sources            []string `json:"sources"`
+	SourceType         *string  `json:"source_type"`
+	Target             *string  `json:"target"`
+	TargetType         *string  `json:"target_type"`
+	PreserveSource     *bool    `json:"preserve_source"`
+	OverrideOnConflict *bool    `json:"override_on_conflict"`
+}
+
+// CategoryProcessor struct represents unique part of category processor object
+// from config API.
+type CategoryProcessor struct {
+	Target     *string    `json:"target"`
+	Categories []Category `json:"categories"`
+}
+
+// Category represents category object from config API.
+type Category struct {
+	Name   *string              `json:"name"`
+	Filter *FilterConfiguration `json:"filter"`
+}
+
+// SourceRemapper represents the object from config API that contains
+// only a list of sources.
+type SourceRemapper struct {
+	Sources []string `json:"sources"`
+}
+
+// GrokParser represents the grok parser processor object from config API.
+type GrokParser struct {
+	Source   *string   `json:"source"`
+	GrokRule *GrokRule `json:"grok"`
+}
+
+// GrokRule represents the rules for grok parser from config API.
+type GrokRule struct {
+	SupportRules *string `json:"support_rules"`
+	MatchRules   *string `json:"match_rules"`
+}
+
+// NestedPipeline represents the pipeline as processor from config API.
+type NestedPipeline struct {
+	Filter     *FilterConfiguration `json:"filter"`
+	Processors []LogsProcessor      `json:"processors,omitempty"`
+}
+
+// UrlParser represents the url parser from config API.
+type UrlParser struct {
+	Sources                []string `json:"sources"`
+	Target                 *string  `json:"target"`
+	NormalizeEndingSlashes *bool    `json:"normalize_ending_slashes"`
+}
+
+// UserAgentParser represents the user agent parser from config API.
+type UserAgentParser struct {
+	Sources   []string `json:"sources"`
+	Target    *string  `json:"target"`
+	IsEncoded *bool    `json:"is_encoded"`
+}
+
+// convert converts the first argument of type interface{} to a map of string and interface{}
+// and sign the result to the second argument.
+func convert(definition interface{}, processor map[string]interface{}) error {
```

Comment:
This method is semi generic, semi specific, e.g. we have generic types with specific argument names, can we make it fully generic or fully specific please ? Also why do we need to marshal unmarshal and not just cast ?

Reply:
I tend to agree that the 3 steps (json marshal/unmarshal and all marshal) seems heavy. Isn't there another way to manage it?